keps/sig-scheduling/3521-pod-scheduling-readiness/README.md

```diff
@@ -965,8 +965,10 @@ Pick one more of these and delete the rest.
 Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
 implementation difficulties, etc.).
 -->
+Yes, we will need a new metric `pod_scheduling_excluding_preenqueue_duration_seconds` to explicitly differentiate whether the time spent during `PreEnqueue` is included or excluded.
```

Comment:
I would prefer we add a new metric that includes the time spent in preenqueue.

The existing metric shouldn't include preenqueue.

cc @ahg-g

Reply:
Precisely. As a stable metric, it's behavior shouldn't change.

IMO, having it consider the PreEnqueue time is a breaking change.

Reply:
> The existing metric shouldn't include preenqueue.

I don't think we should interpret the old metric this way. PreEnqueue falls into the category of scheduler extension point, so once some plugin implements it, it should by default assume its time counted as the e2e scheduling time, and by the new metric, they can opt-in to exclude this extension point explicitly.

Reply:
But the time a Pod spends in PreEnqueue is outside of the control of the scheduler.

The existing metric is meant to be used by SLIs, now we can't use it.

Reply:
> But the time a Pod spends in PreEnqueue is outside of the control of the scheduler.

Arguably this applies to SchedulingGates, but even with that, it's just one usecase of PreEnqueue.

Reply:
That's solvable. There could be a boolean function that PreEnqueue plugins implement to say whether they are externally blocked or not.

Reply:
Are you suggesting a new function for PreEnqueue extension point, and once all PreEnqueue returns "opt-out from e2e duration metric", we go excluding time spent in PreEnqueue?

Reply:
yes, that's one option.

Reply:
Although the only PreEnqueue today is SchedulingGates, so we don't need to backport that function.

Reply:
Think about this case:

- SchedulingGates takes 100s to pass gated check. And, after the gate is lifted, running SchedulingGates is barely a no-op
- PreEnqueue pluginA wants to exclude queued time, same as SchedulingGates
- PreEnqueue pluginB wants to include queued time

So the correct semantics should be: the first time when all "pre-enqueue time opt-out" plugins (in this example, SchedulingGates and pluginA) pass, start the timer. WDYT?

Reply:
I think so. But an extra detail:
There should be validation so that all PreEnqueue plugins that exclude queueing time should be at the beginning of the list.

Reply:
Right, we wouldn't need it.

Reply:
May be if we state the problem, we can brainstorm better solutions?

What we actually want is a metric that measures the scheduler's latency without the impact of external factors. Meaning an SLI for the scheduler itself only. To do this, the SLI should exclude things like node provisioning, volume provisioning and waiting for scheduling gates to be lifted.

The existing `pod_scheduling_duration_seconds` with attempt=0 was our approximation of that (it still includes the volume provisioning overhead for the WaitForFirstConsumer case).

Reply:
> The existing pod_scheduling_duration_seconds with attempt=0 was our approximation of that (it still includes the volume provisioning overhead for the WaitForFirstConsumer case).

pod_scheduling_duration_seconds{attempts=0} won't exist, isn't it? because this metric is generated in [binding](https://github.com/kubernetes/kubernetes/blob/ea1de7517ee5bdffdb31c9a7e5abd6a1fe6b5e0f/pkg/scheduler/schedule_one.go#L261) time, and hence attempts would be definitely >= 1.

Reply:
attempt=1 then.

Reply:
> What we actually want is a metric that measures the scheduler's latency without the impact of external factors.

Yes, a reasonable scheduling latency SLI usually comes with a series of restricting words: like what's defined in [pod_startup_latency.md](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md).

> _By schedulable pod we mean a pod that has to be immediately (without actions from any other components) schedulable in the cluster without causing any preemption._

With that said, scheduler needs to offer building blocks to help customers practically compose metrics to represent scheduling SLIs, and this PR is in-scoped for that.

Back to this topic, if user uses pod_scheduling_duration_seconds{attempts=1} to represent latency SLI, do we agree on proceeding with the semantics mentioned in https://github.com/kubernetes/enhancements/pull/4065#discussion_r1222205444? @ahg-g @alculquicondor @helayoty . (we need to come to a conclusion as KEP deadline is coming)

Reply:
I would prefer we exclude PreEnqueue altogether so that it's easier to backport.

But if you really believe that there will be PreEnqueue plugins that should be considered in the metric (or even better if you have a known use case), then we should only do it for 1.28+, but still backport to 1.27 the simpler fix of excluding all PreEnqueue plugins.

Reply:
> or even better if you have a known use case

let me engage @wilfred-s and @craigcondit as YuniKorn will be a user of PreEnqueue. Also asked in YK's slack channel.

Reply:
To quote @Huang-Wei from the YuniKorn slack conversation:
> The current available options are:
exclude PreEnqueue's time entirely from the e2e metric
add a new e2e metric to exclude PreEnqueue's time
add a function to PreEnqueue extension point to have plugin developers to return true or false to represent they want to exlcude/include the time

I am looking at this from a batch processing view: highly variable loads over time.

Option 1 looks like you are hiding things. In most scheduling systems the overall time of a workload in the system is tracked. Workloads can be queued for some reason. With the K8s scheduler not being aware of quotas and thus no queuing of a workload  the issue has not shown up before now. You really need to have a way to track the overall time without needing to rely on some pod creation time and manual calculations.

Option 3 is a big no go for me. You can not trust the results if certain plugins opt out. What if there is one plugin that wants to opt out and other not? There is no right way to compare and interpret the results without looking at the config either. Even then you know you cannot compare values as you have no split values, released or not by PreEnqueue, in both cases.

I have no problem with changing the current metric to exclude PreEnqueue. Defining that the current metric tracks scheduling time from the point that all the PreEnqueue plugins let it pass. That would be a stable definition that makes it comparable even over releases. It also excludes any external factors from the metric.
Completely excluding always could however be used to hide bottle necks. Not in the scheduler but in the overall system. 

You would want to know how much time the PreEnqueue held up the scheduling. Each plugin is responsible for providing insights into the reason(s) of holding it. From a scheduling perspective you need to at least show knowledge of that overall time.
Time is spend in the PreEnqueue hold. Even if it is managed by an external controller. The submitter wants the pod to run, the cluster is not ready for it yet and PreEnqueue is holding it. Adding a metric which includes the PreEnqueue time provides an insight that would be really valuable.
YuniKorn tracks that as a scheduling delay: waiting for resources to become available to run the workload. When the delay starts to deviate from the normal value that is tracked is an important metric. The change could point to a drop in usage or an overloaded cluster.

Reply:
Thanks @wilfred-s for the input.

> From a scheduling perspective you need to at least show knowledge of that overall time.

We thought of add label to existing metric to divide the overall time by phase (queued vs. ready to schedule), but it's not an option for sig-instrumentation's advise as pod_scheduling_duration_seconds is a stable metric already.

So I think the eventual option can get narrowed down to:

1. Keep `pod_scheduling_duration_seconds` as-is, say including all times spend from pod created to pod bound. In this case, we don't need to backport, but needs to document that this metric is no longer a good fit for SLI. A new metric that excludes preEnqueue is a better option.
2. Change `pod_scheduling_duration_seconds` to exclude the time a Pod spent on preEnqueue. In this case, we need to backport to 1.26 and 1.27, but better confirm with sig-instrumentation beforehand whether it's legal.

With either 1 or 2, we will add a new metric to show e2e time excluding (for 1) / including (for 2) PreEnqueue.

@kubernetes/sig-scheduling-misc

Reply:
The problem with 1 is that we can be breaking existing SLIs in 1.27 and there is no alternate metric that can be used.

Reply:
@helayoty could you help check with @kubernetes/sig-instrumentation-leads if option 2 makes sense with cherrypick PRs to 1.26 and 1.27?

Reply:
2 is legal, a little eye-brow raising, but legal. I would personally probably go with 1, and create a new metric `pod_scheduling_sli_duration_seconds` for the SLI usecase though.

Reply:
Prefer opt-2 still, for:
- We should not **break** user who use `pod_scheduling_duration_seconds` as scheduler latency SLI
- `PreEnqueue` should not be part of the metric since it can be impacted by external factors, not the scheduler itself. And in the foreseeable future, I think `PreEnqueue` will still play a role as **a preparation before scheduling**, hope I understand right here.

Reply:
In addition to the rollback, I think we should still consider adding a new metric that is explicitly focused on the scheduler's own latency, this is more future proof for people who want to create a scheduler latency SLO. `pod_scheduling_sli_duration_seconds` sounds like a good name.

Reply:
Actually I take it back, I'm cool with 2.

Reply:
ok, seems more people are fine with 2, let's go with 2.

I agree with @ahg-g that we need an explicit metric to indicate it's for SLI, at least close to. So is it legal to deprecate `pod_scheduling_duration_seconds` and rename it to `pod_scheduling_sli_duration_seconds`? @logicalhan could you shed some light?

Reply:
I'm ok with deprecating the metric and introducing a new one `pod_scheduling_sli_duration_seconds`, after doing the backport.

The guidance says that stable metrics can be deprecated, but it doesn't say for how long they need to be served before deletion https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/metric-stability.md#stable

Reply:
> The guidance says that stable metrics can be deprecated, but it doesn't say for how long they need to be served before deletion https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/metric-stability.md#stable

[4 releases or 16 months](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-metric)

Reply:
Sounds we reach to a consensus. @helayoty to proceed we can work on the following tasks:

- this KEP will be tweaked a bit to introduce a new metric that is equivalent to what current `pod_scheduling_duration_seconds` behaves - include duration spent on all extension points
- revisit your PR https://github.com/kubernetes/kubernetes/pull/118049 to ensure it's concise enough, as it needs to be back-ported to 1.26 and 1.27
- create a new, duplicate metric (with `pod_scheduling_duration_seconds`) called `pod_scheduling_sli_duration_seconds`, which will explicitly reveals it's a SLI-related metric
- start the deprecation process for `pod_scheduling_duration_seconds`

cc @ahg-g

Reply:
We don't usually fix bugs for alpha features. So the only requirement in the second point is to backport to 1.27, if it makes things easier.

Reply:
KEP has been updated with the new metric.